watchdog: add glue and arm-smc-wdt smc handler

clementleger opened the pull request:
This commits adds a thin glue layer for watchdog access and adds a SMC handler which handles the `arm-smc-wdt` compatible Linux driver [1].

[1] https://lwn.net/Articles/819480/

jenswi-linaro commented on core/drivers/wdt/watchdog.c:
```diff
@@ -0,0 +1,18 @@
+
+struct wdt_chip *wdt_chip;
+
+TEE_Result watchdog_register(struct wdt_chip *chip)
```
No one's calling this function.

jenswi-linaro commented on core/drivers/wdt/watchdog_sm.c:
```diff
@@ -0,0 +1,68 @@
+static unsigned long wdt_min_timeout;
+static unsigned long wdt_max_timeout;
+
+enum sm_handler_ret wdt_sm_handler(struct thread_smc_args *args)
```
No one's calling this function.

clementleger replied:
I will add a commit adding the call to `wdt_sm_handler()`

jenswi-linaro commented on core/drivers/wdt/watchdog_sm.c:
```diff
@@ -0,0 +1,68 @@
+
+enum sm_handler_ret wdt_sm_handler(struct thread_smc_args *args)
+{
+	if (OPTEE_SMC_FUNC_NUM(args->a0) !=
```
The entire id should match, no need to mask out anything.

clementleger replied:
Ack

jenswi-linaro commented on core/drivers/wdt/watchdog_sm.c:
```diff
@@ -0,0 +1,68 @@
+static unsigned long wdt_min_timeout;
+static unsigned long wdt_max_timeout;
+
+enum sm_handler_ret wdt_sm_handler(struct thread_smc_args *args)
```
I'd like to see a definition of the ABI provided by this function. From what I've been able to find so it seems that this is specific for mt8173.

clementleger replied:
Actually, the definition does not seems to be specified by ARM. But the driver that was submitted states that is a generic SMC watchdog driver [1]. If not, were should I place this file ?

[1] https://lore.kernel.org/linux-arm-kernel/20200220155159.GB29658@roeck-us.net/T/

jenswi-linaro replied:
I assume that this might be reused for other platforms. Let's keep it here, but we should be clear on what we're defining.

jenswi-linaro commented on mk/config.mk:
```diff
@@ -789,3 +789,11 @@ $(eval $(call cfg-enable-all-depends,CFG_MEMPOOL_REPORT_LAST_OFFSET, \
+# This watchdog will then be usable by non-secure world through SMC calls.
+CFG_WDT ?= n
+CFG_WDT_SM_HANDLER ?= n
+CFG_WDT_SM_HANDLER_ID ?= 0x82003D06
```
The id seems to be specific to mt8173. It's likely going to be platform specific.

jenswi-linaro replied:
It is quite ugly to be honest. For mt8173 it's the next free SIP function id. Who knows what's suitable for another platform?

clementleger replied:
Agreed, I have switched the config option to the following which forces the user to define it:

```
ifeq ($(CFG_WDT_SM_HANDLER_ID),)
$(error CFG_WDT_SM_HANDLER_ID must be defined when enabling CFG_WDT_SM_HANDLER)
endif
```

jenswi-linaro reviewed (commented):
This code isn't even built in this PR.

clementleger commented:
> This code isn't even built in this PR.

We already did so for the clock framework so I thought it was ok to do it like this. I will add the watchdog driver that uses it and the call `to wdt_sm_handler()` then.

jenswi-linaro commented:
> I will add the watchdog driver that uses it and the call to wdt_sm_handler() then.

Thanks.

clementleger commented:
New version:
- Added call to `wdt_sm_handler()` in core/sm.c
- Added atmel watchdog driver
- Added udelay support without generic timers
- Enabled watchdog on sama5d2

clementleger commented:
New version:
- Fixed merge conflict

clementleger commented:
New version: 
- Fix build when `CFG_WDT_SM_HANDLER_ID` is not defined

jenswi-linaro commented on core/drivers/wdt/watchdog_sm.c:
```diff
@@ -0,0 +1,67 @@
+
+enum sm_handler_ret wdt_sm_handler(struct thread_smc_args *args)
+{
+	if (args->a0 != CFG_WDT_SM_HANDLER_ID)
```
To better optimize the code in `sm_from_nsec()` we could do this test in an inline function which would then be executed as part of `sm_from_nsec()` avoiding the jump to this function only to return most often. Do you get the idea?

clementleger replied:
Yes totally. This would avoid a function call each time we do the test, acked.

clementleger commented:
New version:
- Apply Jens recommendation to do the SMC ID check in an inlined function

jenswi-linaro commented on core/arch/arm/kernel/delay.c:
```diff
@@ -27,15 +27,30 @@
+	uint64_t cycles = 0;
+
+	cycles = (uint64_t)us * ((uint64_t)plat_get_freq() / 1000000ULL);
+	assert(cycles < (unsigned long)-1);
```
What is the purpose of this assert? When do we care if there's an overflow or not?

clementleger replied:
Since the `udelay()` is implemented in assembly, it uses 32bits registers and I did not wanted to handle 32/64bits in assembly code, that's why I added this assert here.

jenswi-linaro replied:
Using an assert means that you don't care about this in production builds, only during development.
Perhaps it would be better to just set a max value instead in case of overflow.

clementleger replied:
Yes, sorry, indeed, using a maximum value would actually be better.

clementleger replied:
Actually, what can be done is simply to loop on the `wait_cycles` function by calling it with `MIN(cycles, UINT32T_MAX)` and this will handle the 64 bits values.

jenswi-linaro replied:
Does it make sense to spin for such a long time? If we're close to UINT32_MAX clearly something is wrong.

clementleger replied:
Actually, the `uint32_t` is meant to hold the number of cycles which might be way greater than the number of usec.

jenswi-linaro replied:
OK, got it. Looping as you suggest makes sense.

jenswi-linaro commented on core/arch/arm/plat-sam/freq.c:
```diff
@@ -0,0 +1,39 @@
+
+static unsigned long freq;
+
+static unsigned long get_freq_from_dt(void)
```
I'd prefer doing this once at boot so we can forget about the embedded DT once boot is done.

clementleger replied:
Since I'm using the clock framework, it is not ready at boot time but just after the clock framework is probed. Doing so allows to defer the clock retrieval.

jenswi-linaro replied:
The point is that I'd like to avoid depending on an embedded DT once boot is done. Let's say that the embedded DT is replaced by a DT which is only available until we do the first exit to normal world.

clementleger replied:
Hum ok, I understand. So indeed, I will try to add an early initicall to retrieve the frequency from the DT with the clock framework once and thus setup the plat_freq variable at init only.

jenswi-linaro commented on core/drivers/atmel_wdt.c:
```diff
@@ -0,0 +1,222 @@
+static void atmel_wdt_write_sleep(struct atmel_wdt *wdt, uint32_t reg,
+				  uint32_t val)
+{
+	udelay(WDT_REG_ACCESS_UDELAY);
```
You should perhaps call `wait_cycles()` instead with the number of cycles you'd like to wait.

clementleger replied:
Hum, wouldn't this be breaking the fact that the driver can work on multiple platforms ? That why I added the udelay layer that uses the platform frequency.

jenswi-linaro replied:
Sorry, my mistake.

jenswi-linaro commented on core/drivers/wdt/watchdog_sm.c:
```diff
@@ -0,0 +1,64 @@
+
+enum sm_handler_ret __wdt_sm_handler(struct thread_smc_args *args)
+{
+	switch (args->a1) {
```
What happens if this function is called concurrently on two different CPUs?

clementleger replied:
Main concern about that is `wdt_min_timeout` and `wdt_max_timeout` values that are potentially accessed concurrently even if the same values will be returned. I'm not sure where the concurrency should be handled. From what I already saw in OP-TEE, it seems you prefer to handle this in drivers but I'm not entirely sure.

jenswi-linaro replied:
`wdt_min_timeout` and `wdt_max_timeout` must be protected here since they are read in this function. Other aspects should probably be handled inside the driver in case that driver is accessed somewhere else too.

clementleger commented on core/include/drivers/wdt.h:
```diff
@@ -27,10 +29,76 @@ struct wdt_chip {
+extern struct wdt_chip *wdt_chip;
+
+/* Register a watchdog as the system watchdog */
+TEE_Result watchdog_register(struct wdt_chip *chip);
```
I will actually squash the two first commits il does not makes a lot of sens to register a driver but not being able to really use it with a proper API.

clementleger commented:
New version:
- Allow to handle large cycles count
- Read freq at boot time and not when calling `plat_get_freq()`
- Squashed missing suspend support for sama5d2 watchdog in watchdog driver

clementleger commented:
New version:
- Add locking for min/max timeout variables

jenswi-linaro commented:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

clementleger commented:
BTW, do you think the `atmel_wdt` driver should be moved in drivers/wdt ? There are already some watchdog drivers in drivers/ but maybe it's time to start organizing watchdog drivers in drivers/wdt/

clementleger commented:
New version:
- Added Jens Acked-by
- Fixed conflicts

clementleger commented:
Finally just replaced `assert(wdt_chip)` in watchdog interface with `if (!wdt_chip) return;`. Without that, if compiled without debug, then the normal world could potentially crash OP-TEE by calling `wdt_start()` or others when no watchdog device is present. This does not seems a good things to allow.

jenswi-linaro commented on core/include/drivers/wdt.h:
```diff
@@ -27,10 +31,99 @@ struct wdt_chip {
+	if (!wdt_chip)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	if (wdt_chip->ops->init)
```
How about combining the check above with this to become
``` C
if (wdt_chip && wdt_chip->ops->init)
```
It saves a few lines looks better in my opinion. Same in for instance `watchdog_stop()` below.

clementleger replied:
Acked, actually lack a `init` function should not be fatal and we should simply return `TEE_SUCCESS`

jenswi-linaro commented on core/include/drivers/wdt.h:
```diff
@@ -27,10 +31,99 @@ struct wdt_chip {
+	if (!wdt_chip)
+		return;
+
+	wdt_chip->ops->start(wdt_chip);
```
How about
``` C
if (wdt_chip)
         wdt_chip->ops->start(wdt_chip);
```
instead? Same in `watchdog_ping()` and `watchdog_settimeout()` below.

clementleger replied:
Agreed

clementleger commented:
New version:
- Simplify ops check for readability

jenswi-linaro commented:
Looks good.
